### ISSUE TITLE
Add a rule which disallows calling a non-void/non-never function without using its return value

### ISSUE THREAD

**Issue by dko-slapdash** [typescript-eslint/typescript-eslint#1314]
In C++/gcc, there is a useful attribute which can be assigned to a function, and then swallowing the returned value of this function will be considered as an error: `__attribute__((warn_unused_result))`. 

Is there something like that for eslint/typescript?

**What code were you trying to parse?**

```ts
// eslint-enable-next-line @typescript-eslint/no_unused_result <-- ???
function f() {
  return 10;
}

f();
```

**What did you expect to happen?**

It should warn that the function's return value is not consumed and is swallowed. There is also no eslint-enable-next-line (but maybe there is something instead?).

**What actually happened?**

Nothing, I couldn't find such a rule. Surprisingly, no-unused-expressions doesn't help here, it doesn't process function returns.

**Comment by bradzacher** [comment]
> Surprisingly, no-unused-expressions doesn't help here, it doesn't process function returns.

Function return values are not expressions. Function calls are expressions.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#Expressions

> An expression is any valid unit of code that resolves to a value

---

There is ***no*** rule to check if a function with a non-void return value is being used without storing its return value.

Happy to accept a PR to add one.

```ts
declare function f(): Promise<number> | number;

// valid
const x = f();
function g() {
  return f();
}
void f();
await f();
typeof f() === 'number';
// etc

// invalid
f();
```

**Comment by robyoder** [comment]
This would be lovely. I was searching for this because I'm refactoring some code to return things that used to return void or Promise<void>. I'll add it to my list to potentially cut my teeth on here if I can. :)

I'd disagree on `await f();` being valid, but I suppose that could be an option in the rule. Personally, I would only want that to be valid if the return type was `Promise<void>` because many of the functions I'm working with are returning Promises, so it would defeat the purpose if you awaited and then did nothing with the awaited value.

**Comment by JoshuaKGoldberg** [comment]
This rule would have a _ton_ of false negatives. Many functions return a value that indicates some result or state change even if it's not relevant to the caller. I'm not sold that we want this as part of typescript-eslint core.

One example is `Array`'s `.push` returning the new length of the array.

```ts
const items = [];
items.push('a', 'b', 'c'); // 3
```

cc @bradzacher - I'm inclined to close this issue as one that should be implemented externally.

**Comment by bradzacher** [comment]
yeah definitely - it's really common to have functions which return ignorable values.
another example is the "builder" pattern for which methods all return `this`.

I don't think this is a good candidate for a lint rule as it would be too noisy.
